File: app/src/main/java/in/completecourse/fragment/mainFragment/HomeFragment.java
package in.completecourse.fragment.mainFragment;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

import in.completecourse.R;
import in.completecourse.ScanActivity;
import in.completecourse.SearchActivity;
import in.completecourse.SubjectActivity;
import in.completecourse.adapter.ImageAdapter;
import in.completecourse.adapter.SliderAdapter;
import in.completecourse.app.AppConfig;
import in.completecourse.helper.HelperMethods;
import in.completecourse.helper.HttpHandler;
import in.completecourse.model.CardModel;
import in.completecourse.model.Update;


public class HomeFragment extends Fragment {

    private ViewPager mViewPager;
    private ArrayList<Update> updateList;
    private static String urlQR;
    private SliderAdapter sliderAdapter;

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        //super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 100 && resultCode == 100 && data != null){
            urlQR = data.getStringExtra("url");
            new GetBookName(HomeFragment.this).execute();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        ArrayList<CardModel> cardList = new ArrayList<>();
        cardList.add(new CardModel(R.drawable.manual_search, "Manual Search"));
        cardList.add(new CardModel(R.drawable.scan_qr, "Scan QR Code"));
        mViewPager = view.findViewById(R.id.viewPager);
        TabLayout indicator = view.findViewById(R.id.indicator);
        RecyclerView recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);
        //use a linear layout manager
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext() ,2, RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(gridLayoutManager);

        //specify an adapter
        ImageAdapter recyclerViewAdapter = new ImageAdapter(cardList, getContext());
        recyclerView.setAdapter(recyclerViewAdapter);
        recyclerView.addOnItemTouchListener(new ImageAdapter.RecyclerTouchListener(getContext(), position -> {
            switch (position){
                case 0:
                    Intent searchActivityIntent = new Intent(getContext(), SearchActivity.class);
                    startActivity(searchActivityIntent);
                    break;
                case 1:
                    Intent qrCodeActivityIntent = new Intent(getContext(), ScanActivity.class);
                    startActivityForResult(qrCodeActivityIntent, 100);
                    break;
            }

        }));

        updateList = new ArrayList<>();
        sliderAdapter = new SliderAdapter(getActivity(), updateList);
        mViewPager.setAdapter(sliderAdapter);
        indicator.setupWithViewPager(mViewPager, true);
        Timer timer = new Timer();
        timer.scheduleAtFixedRate(new SliderTimer(), 4000, 6000);

        if (HelperMethods.isNetworkAvailable(getActivity())){

            JSONTransmitter jsonTransmitter = new JSONTransmitter(HomeFragment.this);
            jsonTransmitter.execute();

        }else{
            Toast.makeText(view.getContext(), "Please check your internet connection.", Toast.LENGTH_SHORT).show();
        }
    }

    private class SliderTimer extends TimerTask {
        @Override
        public void run() {
            if(getActivity() != null){
                getActivity().runOnUiThread(() -> {
                    if (mViewPager.getCurrentItem() < updateList.size() - 1) {
                        mViewPager.setCurrentItem(mViewPager.getCurrentItem() + 1);
                    } else {
                        mViewPager.setCurrentItem(0);
                    }
                });
            }
        }
    }

    private static class GetBookName extends AsyncTask<Void, Void, Void> {
        private final WeakReference<HomeFragment> activityWeakReference;

        GetBookName(HomeFragment context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected Void doInBackground(Void... arg0) {
            final HomeFragment activity = activityWeakReference.get();
            HttpHandler sh = new HttpHandler();
            String url = urlQR;
            String jsonStr = sh.makeServiceCall(url);

            if (jsonStr != null) {
                try {
                    JSONObject jsonObject = new JSONObject(jsonStr);
                    String studentSubject = jsonObject.getString("studentsubject");
                    String studentClass = jsonObject.getString("studentclass");
                    Intent intent = new Intent(activity.getActivity(), SubjectActivity.class);
                    intent.putExtra("classCode", studentClass);
                    intent.putExtra("subjectCode", studentSubject);
                    activity.startActivity(intent);
                }catch (final JSONException e) {
                    if (activity.getActivity() != null) {
                        activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(),
                                "Json parsing error: " + e.getMessage(),
                                Toast.LENGTH_LONG)
                                .show());
                    }

                }
            } else {
                if (activity.getActivity() != null) {
                    activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), "Couldn't get json from server. Check LogCat for possible errors!",
                            Toast.LENGTH_LONG)
                            .show());
                }

            }

            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);

        }

    }

    static class JSONTransmitter extends AsyncTask<String, String, String> {
        private final WeakReference<HomeFragment> activityWeakReference;

        JSONTransmitter(HomeFragment context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected void onPreExecute() {
            HomeFragment activity = activityWeakReference.get();
        }

        @Override
        protected String doInBackground(String... params) {
            final HomeFragment activity = activityWeakReference.get();
            String urlString = AppConfig.URL_UPDATES;
            URL url;
            InputStream stream;
            HttpURLConnection urlConnection = null;
            try {
                url = new URL(urlString);
                urlConnection = (HttpURLConnection) url.openConnection();
                urlConnection.setRequestMethod("GET");
                urlConnection.setDoOutput(true);
                urlConnection.connect();
                stream = urlConnection.getInputStream();
                BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8), 8);

                String resFromServer = reader.readLine();
                String status;
                JSONObject jsonResponse;
                try {
                    jsonResponse = new JSONObject(resFromServer);
                    //Log.e("chatpterItem", String.valueOf(jsonResponse));
                    status = jsonResponse.getString("status");
                    if (status.equals("true")){
                        JSONObject jsonObject = new JSONObject(resFromServer);
                        JSONArray jsonArray = jsonObject.getJSONArray("data");
                        Update update;
                        for (int i = 0; i < jsonArray.length(); i++){
                            JSONObject dataObject = jsonArray.getJSONObject(i);
                            update = new Update(dataObject.getString("name"), dataObject.getString("imageurl"));
                            activity.updateList.add(update);
                            activity.sliderAdapter.setItems(activity.updateList);
                            if (activity.getActivity() != null) {
                                activity.getActivity().runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        activity.sliderAdapter.notifyDataSetChanged();
                                    }
                                });
                            }
                        }

                    }else{
                        final String msg = jsonResponse.getString("message");
                        if (activity.getActivity() != null) {
                            activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), msg, Toast.LENGTH_SHORT).show());
                        }
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                return null; //reader.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                if (urlConnection != null) {
                    urlConnection.disconnect();
                }
            }
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                e.printStackTrace();
                Log.i("Result", "SLEEP ERROR");
            }
            return null;
        }

        @Override
        protected void onPostExecute(String jsonObject) {

        }
    }
}

File: app/src/main/java/in/completecourse/fragment/mainFragment/HomeFragment.kt
package `in`.completecourse.fragment.mainFragment

import `in`.completecourse.R
import `in`.completecourse.ScanActivity
import `in`.completecourse.SearchActivity
import `in`.completecourse.SubjectActivity
import `in`.completecourse.adapter.ImageAdapter
import `in`.completecourse.adapter.SliderAdapter
import `in`.completecourse.databinding.FragmentHomeBinding
import `in`.completecourse.helper.HelperMethods
import `in`.completecourse.model.CardModel
import `in`.completecourse.model.Update
import `in`.completecourse.utils.APIService
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import retrofit2.Retrofit
import java.util.*

class HomeFragment : Fragment(), ImageAdapter.ClickListener {
    private var updateList: ArrayList<Update>? = null
    private var sliderAdapter: SliderAdapter? = null

    private var _binding: FragmentHomeBinding? = null
    //This property is only valid between onCreateView and
    //onDestroyView
    private val binding get() = _binding!!

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        //super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 100 && resultCode == 100 && data != null) {
            urlQR = data.getStringExtra("url")
            if (urlQR != null) {
                val intent = Intent(context, SubjectActivity::class.java)
                when (urlQR) {
                    "http://completecourse.in/api/GetQRValue/4/1" -> {
                        intent.putExtra("classCode", "4")
                        intent.putExtra("subjectCode", "1")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/4/4" -> {
                        intent.putExtra("classCode", "4")
                        intent.putExtra("subjectCode", "4")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/1/2" -> {
                        intent.putExtra("classCode", "1")
                        intent.putExtra("subjectCode", "2")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/1/3" -> {
                        intent.putExtra("classCode", "1")
                        intent.putExtra("subjectCode", "3")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/2/7" -> {
                        intent.putExtra("classCode", "2")
                        intent.putExtra("subjectCode", "7")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/2/9" -> {
                        intent.putExtra("classCode", "2")
                        intent.putExtra("subjectCode", "9")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/2/11" -> {
                        intent.putExtra("classCode", "2")
                        intent.putExtra("subjectCode", "11")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/2/5" -> {
                        intent.putExtra("classCode", "2")
                        intent.putExtra("subjectCode", "5")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/3/8" -> {
                        intent.putExtra("classCode", "3")
                        intent.putExtra("subjectCode", "8")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/3/10" -> {
                        intent.putExtra("classCode", "3")
                        intent.putExtra("subjectCode", "10")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/3/12" -> {
                        intent.putExtra("classCode", "3")
                        intent.putExtra("subjectCode", "12")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/3/6" -> {
                        intent.putExtra("classCode", "3")
                        intent.putExtra("subjectCode", "6")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/4/16" -> {
                        intent.putExtra("classCode", "4")
                        intent.putExtra("subjectCode", "16")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/4/15" -> {
                        intent.putExtra("classCode", "4")
                        intent.putExtra("subjectCode", "15")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/1/13" -> {
                        intent.putExtra("classCode", "1")
                        intent.putExtra("subjectCode", "15")
                        startActivity(intent)
                    }
                    "http://completecourse.in/api/GetQRValue/1/14" -> {
                        intent.putExtra("classCode", "1")
                        intent.putExtra("subjectCode", "14")
                        startActivity(intent)
                    }
                }
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        val cardList: ArrayList<CardModel> = ArrayList<CardModel>()
        cardList.add(CardModel(R.drawable.manual_search, "Manual Search"))
        cardList.add(CardModel(R.drawable.scan_qr, "Scan QR Code"))

        binding.recyclerView.setHasFixedSize(true)
        //use a linear layout manager
        val gridLayoutManager = GridLayoutManager(context, 2, RecyclerView.VERTICAL, false)
        binding.recyclerView.layoutManager = gridLayoutManager

        //specify an adapter
        val recyclerViewAdapter = ImageAdapter(cardList, context!!)
        binding.recyclerView.adapter = recyclerViewAdapter
        binding.recyclerView.addOnItemTouchListener(ImageAdapter.RecyclerTouchListener(context, this))

        val timer = Timer()
        timer.scheduleAtFixedRate(SliderTimer(), 4000, 6000)
        if (HelperMethods.isNetworkAvailable(activity)) getUpdates()
        else Toast.makeText(view.context, "Please check your internet connection.", Toast.LENGTH_SHORT).show()
    }

    private inner class SliderTimer : TimerTask() {
        override fun run() {
            if (activity != null) {
                activity!!.runOnUiThread {
                    if (binding.viewPager.currentItem < updateList?.size?.minus(1) ?: 0) {
                        binding.viewPager.currentItem = binding.viewPager.currentItem + 1
                    } else {
                        binding.viewPager.currentItem = 0
                    }
                }
            }
        }
    }

    private fun getUpdates() {
        //binding.progressBar.visibility = View.VISIBLE
        //Create Retrofit
        val retrofit = Retrofit.Builder().baseUrl("http://completecourse.in/api/").build()

        //Create Service
        val service = retrofit.create(APIService::class.java)

        CoroutineScope(Dispatchers.IO).launch {
            //Do the POST request and get the response
            val response = service.getUpdates()
            if (response.isSuccessful) {
                var update: Update
                val obj = JSONObject(response.body()?.string() ?: "{}")
                val jsonArray = obj.getJSONArray("data")//JSONArray(obj.getJSONArray("data"))
                //Log.e("jsonArray", jsonArray.toString())
                for (i in 0 until jsonArray.length()) {
                    val dataObject = jsonArray.getJSONObject(i)
                    update = Update(dataObject.getString("name"), dataObject.getString("imageurl"))
                    updateList = ArrayList()
                    updateList?.add(update)
                }
            }

            withContext(Dispatchers.Main) {
                sliderAdapter = SliderAdapter(activity, updateList!!)
                sliderAdapter?.setItems(updateList)
                binding.indicator.setupWithViewPager(binding.viewPager, true)
                binding.viewPager.adapter = sliderAdapter
            }
        }
    }

    companion object {
        private var urlQR: String? = null
    }

    override fun onClick(position: Int) {
        when (position) {
            0 -> {
                val searchActivityIntent = Intent(context, SearchActivity::class.java)
                startActivity(searchActivityIntent)
            }
            1 -> {
                val qrCodeActivityIntent = Intent(context, ScanActivity::class.java)
                startActivityForResult(qrCodeActivityIntent, 100)
            }
        }
    }
}